# James Velardi

## April 27th, 2025

## Final Project Milestone #3

## Data Cleaning

### Purpose of This Project

<p>This project will investigate vehicle safety data from NHTSA in 2024. Data Science questions will be asked and plotted in determining which cars in 2011-2024 are the safest for consumers. There’s a lot of potential clients who could use this data such as consumers, rental car companies, automobile insurance, etc. For the sake of this project, this project will aim more towards rental car companies, seeing as how important safety is to these companies. Overall, the NHTSA has plenty of good data to use. The plan will be to clean and filter it down to a usable level, as the dataset has a lot of data stored inside of it. </p>

### Data Science Questions to Explore

1.	Which car manufacturers overall have the highest average 5-star safety ratings. 
2.	What is the distribution of vehicle types (Sedan, SUV, etc.) and their safety ratings. 
3.	Do cars with advanced safety features (Automatic Emergency Braking, Forward Collision Warning) help improve safety ratings? 
4.	How does the safety of vehicles compare overtime? Are newer vehicles getting much safer?
5.	What do vehicles that rank the lowest in vehicle safety have the most in common? Could it be based on body type, frame type, lack of safety features and airbags, or does it all come down to brand? 


### Import Pandas

In [32]:
import pandas as pd

import re

import numpy as np

print(f'Regex Version: {re.__version__}')

print(f'Pandas Version: {pd.__version__}')

print(f'Numpy Version: {np.__version__}')


Regex Version: 2.2.1
Pandas Version: 2.2.3
Numpy Version: 1.26.4


### Turn the CSV file into a dataframe

In [2]:
data = 'Safercar_data.csv'

df = pd.read_csv(data, low_memory=False)

df

MAKE           MODEL  MODEL_YR BODY_STYLE VEHICLE_TYPE DRIVE_TRAIN  \
0      ACURA         INTEGRA      2024       5 HB           PC         FWD   
1      ACURA  INTEGRA TYPE-S      2024       5 HB           PC         FWD   
2      ACURA             MDX      2024        SUV          MPV         AWD   
3      ACURA             MDX      2024        SUV          MPV         FWD   
4      ACURA             RDX      2024        SUV          MPV         AWD   
...      ...             ...       ...        ...          ...         ...   
11023  VOLVO            XC60      2011        SUV          MPV         FWD   
11024  VOLVO            XC70      2011         SW          MPV         AWD   
11025  VOLVO            XC70      2011         SW          MPV         FWD   
11026  VOLVO            XC90      2011        SUV          MPV         AWD   
11027  VOLVO            XC90      2011        SUV          MPV         FWD   

       PRODUCTION_RELEASE VEHICLE_CLASS BODY_FRAME NUM_OF_SEATING  ...  \
0                       1           NaN                         5  ...   
1                       1           NaN                         4  ...   
2                       1           NaN                         7  ...   
3                       1           NaN                         7  ...   
4                       1           NaN                         5  ...   
...                   ...           ...        ...            ...  ...   
11023                   1           SUV   Uni-Body              5  ...   
11024                   1           SUV   Uni-Body              5  ...   
11025                   1           SUV   Uni-Body              5  ...   
11026                   1           SUV   Uni-Body              7  ...   
11027                   1           SUV   Uni-Body              7  ...   

      POLE_HIC_36_DRIV PELVIC_FORCE  ROLLOVER_POSSIBILITY  \
0              235.989     2871.970                 0.095   
1                  NaN          NaN                 0.095   
2              549.430     1886.689                 0.151   
3              549.430     1886.689                 0.151   
4              485.973     3130.952                 0.165   
...                ...          ...                   ...   
11023          242.120     3207.978                 0.176   
11024              NaN          NaN                 0.000   
11025              NaN          NaN                 0.000   
11026              NaN          NaN                 0.179   
11027              NaN          NaN                 0.191   

       STATIC_STABI_FACTOR     TIP ROLL_SAFETY_CONCERN ROLL_FOOT_NOTES  \
0                   1.4800  No Tip                 NaN             NaN   
1                   1.4800  No Tip                 NaN             NaN   
2                      NaN  No Tip                 NaN             NaN   
3                   1.2700  No Tip                 NaN             NaN   
4                   1.2372  No Tip                 NaN             NaN   
...                    ...     ...                 ...             ...   
11023               1.2160  No Tip                 NaN             NaN   
11024               0.0000     NaN                 NaN             NaN   
11025               0.0000     NaN                 NaN             NaN   
11026               1.2100  No Tip                 NaN             NaN   
11027               1.1900  No Tip                 NaN             NaN   

      ROLLOVER_STARS NHTSA_BACKUP_CAMERA BACKUP_CAMERA  
0                5.0                 NaN           NaN  
1                5.0                 NaN           NaN  
2                NaN                 NaN           NaN  
3                4.0                 NaN           NaN  
4                4.0                 NaN           NaN  
...              ...                 ...           ...  
11023            4.0                 NaN           NaN  
11024            NaN                 NaN           NaN  
11025            NaN                 NaN       

### Run the describe function

In [3]:
df.describe()

MODEL_YR  PRODUCTION_RELEASE  MIN_GROSS_WEIGHT  MAX_GROSS_WEIGHT  \
count  11028.000000        11028.000000      10002.000000      10236.000000   
mean    2017.763783            1.010247       5614.224575       5917.763599   
std        4.004007            0.102495       1673.910034       1907.128440   
min     2011.000000            1.000000          7.000000        440.000000   
25%     2014.000000            1.000000       4471.250000       4696.000000   
50%     2018.000000            1.000000       5176.000000       5455.000000   
75%     2021.000000            1.000000       6393.000000       6600.000000   
max     2024.000000            3.000000      13000.000000      14000.000000   

       OVERALL_STARS  FRNT_TEST_NO  FRNT_DRIV_STARS  FRNT_PASS_STARS  \
count    4670.000000   4976.000000      4981.000000      4978.000000   
mean        4.562313   9455.062098         4.455129         4.202089   
std         0.576428   1791.361638         0.638022         0.740989   
min         2.000000   6940.000000         2.000000         2.000000   
25%         4.000000   7975.000000         4.000000         4.000000   
50%         5.000000   9248.000000         5.000000         4.000000   
75%         5.000000  10704.000000         5.000000         5.000000   
max         5.000000  14433.000000         5.000000         5.000000   

       OVERALL_FRNT_STARS  CURB_WEIGHT  ...  SYMPHYSIS_FORCE_DRIV  \
count         4978.000000  5411.000000  ...           4557.000000   
mean             4.283849  4093.355202  ...           1402.657576   
std              0.620771  1169.349662  ...            528.287458   
min              2.000000     0.000000  ...            481.118000   
25%              4.000000  3452.000000  ...           1035.229000   
50%              4.000000  3989.000000  ...           1366.163000   
75%              5.000000  4679.000000  ...           1696.713000   
max              5.000000  7586.000000  ...           3969.616000   

       SIDE_HIC_36_PASS  PELVIC_FORCE_PASS  POLE_TEST_NO  SIDE_POLE_STARS  \
count       4411.000000        4411.000000   4976.000000      4983.000000   
mean         192.419793        2558.703694   9409.448352         4.798114   
std          134.216494        1054.450987   1722.508558         0.632154   
min            9.817000         245.903000   6942.000000         1.000000   
25%           88.700000        1775.072000   8070.000000         5.000000   
50%          162.245000        2547.281000   9083.500000         5.000000   
75%          273.763000        3296.632000  10650.000000         5.000000   
max          819.636000        6357.052000  14460.000000         5.000000   

       POLE_HIC_36_DRIV  PELVIC_FORCE  ROLLOVER_POSSIBILITY  \
count       4568.000000   4568.000000          11028.000000   
mean         353.336207   3225.864553              0.091615   
std          717.159456    796.345045              0.103418   
min           85.384000   1393.511000              0.000000   
25%          239.029000   2653.910000              0.000000   
50%          294.402000   3144.494000              0.095000   
75%          364.533000   3713.431000              0.164000   
max        12246.016000   7430.367000              1.330000   

       STATIC_STABI_FACTOR  ROLLOVER_STARS  
count         10776.000000     6118.000000  
mean              0.732125        3.984962  
std               0.648867        0.688851  
min               0.000000        1.000000  
25%               0.000000        4.000000  
50%               1.150000        4.000000  
75%               1.270000        4.000000  
max               1.830000        5.000000  

[8 rows x 44 columns]

### Get the columns

In [4]:
df.columns

Index(['MAKE', 'MODEL', 'MODEL_YR', 'BODY_STYLE', 'VEHICLE_TYPE',
       'DRIVE_TRAIN', 'PRODUCTION_RELEASE', 'VEHICLE_CLASS', 'BODY_FRAME',
       'NUM_OF_SEATING',
       ...
       'POLE_HIC_36_DRIV', 'PELVIC_FORCE', 'ROLLOVER_POSSIBILITY',
       'STATIC_STABI_FACTOR', 'TIP', 'ROLL_SAFETY_CONCERN', 'ROLL_FOOT_NOTES',
       'ROLLOVER_STARS', 'NHTSA_BACKUP_CAMERA', 'BACKUP_CAMERA'],
      dtype='object', length=128)

### Get the amount of rows and columns

In [5]:
df.shape

(11028, 128)

### Get the data types

In [6]:
df.dtypes

MAKE                    object
MODEL                   object
MODEL_YR                 int64
BODY_STYLE              object
VEHICLE_TYPE            object
                        ...   
ROLL_SAFETY_CONCERN     object
ROLL_FOOT_NOTES         object
ROLLOVER_STARS         float64
NHTSA_BACKUP_CAMERA     object
BACKUP_CAMERA           object
Length: 128, dtype: object

### Get the first few rows in the dataframe

In [7]:
df.head()

MAKE           MODEL  MODEL_YR BODY_STYLE VEHICLE_TYPE DRIVE_TRAIN  \
0  ACURA         INTEGRA      2024       5 HB           PC         FWD   
1  ACURA  INTEGRA TYPE-S      2024       5 HB           PC         FWD   
2  ACURA             MDX      2024        SUV          MPV         AWD   
3  ACURA             MDX      2024        SUV          MPV         FWD   
4  ACURA             RDX      2024        SUV          MPV         AWD   

   PRODUCTION_RELEASE VEHICLE_CLASS BODY_FRAME NUM_OF_SEATING  ...  \
0                   1           NaN                         5  ...   
1                   1           NaN                         4  ...   
2                   1           NaN                         7  ...   
3                   1           NaN                         7  ...   
4                   1           NaN                         5  ...   

  POLE_HIC_36_DRIV PELVIC_FORCE  ROLLOVER_POSSIBILITY  STATIC_STABI_FACTOR  \
0          235.989     2871.970                 0.095               1.4800   
1              NaN          NaN                 0.095               1.4800   
2          549.430     1886.689                 0.151                  NaN   
3          549.430     1886.689                 0.151               1.2700   
4          485.973     3130.952                 0.165               1.2372   

      TIP ROLL_SAFETY_CONCERN ROLL_FOOT_NOTES ROLLOVER_STARS  \
0  No Tip                 NaN             NaN            5.0   
1  No Tip                 NaN             NaN            5.0   
2  No Tip                 NaN             NaN            NaN   
3  No Tip                 NaN             NaN            4.0   
4  No Tip                 NaN             NaN            4.0   

  NHTSA_BACKUP_CAMERA BACKUP_CAMERA  
0                 NaN           NaN  
1                 NaN           NaN  
2                 NaN           NaN  
3                 NaN           NaN  
4                 NaN           NaN  

[5 rows x 128 columns]

### Get the last few rows in the dataframe

In [8]:
df.tail()

MAKE MODEL  MODEL_YR BODY_STYLE VEHICLE_TYPE DRIVE_TRAIN  \
11023  VOLVO  XC60      2011        SUV          MPV         FWD   
11024  VOLVO  XC70      2011         SW          MPV         AWD   
11025  VOLVO  XC70      2011         SW          MPV         FWD   
11026  VOLVO  XC90      2011        SUV          MPV         AWD   
11027  VOLVO  XC90      2011        SUV          MPV         FWD   

       PRODUCTION_RELEASE VEHICLE_CLASS BODY_FRAME NUM_OF_SEATING  ...  \
11023                   1           SUV   Uni-Body              5  ...   
11024                   1           SUV   Uni-Body              5  ...   
11025                   1           SUV   Uni-Body              5  ...   
11026                   1           SUV   Uni-Body              7  ...   
11027                   1           SUV   Uni-Body              7  ...   

      POLE_HIC_36_DRIV PELVIC_FORCE  ROLLOVER_POSSIBILITY  \
11023           242.12     3207.978                 0.176   
11024              NaN          NaN                 0.000   
11025              NaN          NaN                 0.000   
11026              NaN          NaN                 0.179   
11027              NaN          NaN                 0.191   

       STATIC_STABI_FACTOR     TIP ROLL_SAFETY_CONCERN ROLL_FOOT_NOTES  \
11023                1.216  No Tip                 NaN             NaN   
11024                0.000     NaN                 NaN             NaN   
11025                0.000     NaN                 NaN             NaN   
11026                1.210  No Tip                 NaN             NaN   
11027                1.190  No Tip                 NaN             NaN   

      ROLLOVER_STARS NHTSA_BACKUP_CAMERA BACKUP_CAMERA  
11023            4.0                 NaN           NaN  
11024            NaN                 NaN           NaN  
11025            NaN                 NaN           NaN  
11026            4.0                 NaN           NaN  
11027            4.0                 NaN           NaN  

[5 rows x 128 columns]

### Get a random row in the dataframe

In [9]:
df.sample()

MAKE           MODEL  MODEL_YR BODY_STYLE VEHICLE_TYPE DRIVE_TRAIN  \
4037  PORSCHE  718 CAYMAN GTS      2020       2 DR           PC         RWD   

      PRODUCTION_RELEASE VEHICLE_CLASS BODY_FRAME NUM_OF_SEATING  ...  \
4037                   1           NaN   Uni-Body              2  ...   

     POLE_HIC_36_DRIV PELVIC_FORCE  ROLLOVER_POSSIBILITY  STATIC_STABI_FACTOR  \
4037              NaN          NaN                   0.0                  0.0   

      TIP ROLL_SAFETY_CONCERN ROLL_FOOT_NOTES ROLLOVER_STARS  \
4037  NaN                 NaN             NaN            NaN   

     NHTSA_BACKUP_CAMERA BACKUP_CAMERA  
4037                 NaN           NaN  

[1 rows x 128 columns]

### Get a quick summary of the dataframe

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11028 entries, 0 to 11027
Columns: 128 entries, MAKE to BACKUP_CAMERA
dtypes: float64(42), int64(2), object(84)
memory usage: 10.8+ MB


### Observations So Far

<p>Upon initial viewing of the data, I noticed in the describe function that newer vehicles have higher overall 5-star ratings.</p>

<p>Since there is a lot of data in this dataset, I am going to filter out the columns that are not needed to answer the data science questions. </p>

In [11]:
data_df = df[['MAKE','MODEL_YR','VEHICLE_TYPE','DRIVE_TRAIN','BODY_FRAME','BLIND_SPOT_DETECTION','ARS','FRNT_COLLISION_WARNING','LANE_DEPARTURE_WARNING','CRASH_IMMINENT_BRAKE','DYNAMIC_BRAKE_SUPPORT','OVERALL_STARS','BACKUP_CAMERA']]
data_df

MAKE  MODEL_YR VEHICLE_TYPE DRIVE_TRAIN BODY_FRAME  \
0      ACURA      2024           PC         FWD              
1      ACURA      2024           PC         FWD              
2      ACURA      2024          MPV         AWD              
3      ACURA      2024          MPV         FWD              
4      ACURA      2024          MPV         AWD              
...      ...       ...          ...         ...        ...   
11023  VOLVO      2011          MPV         FWD   Uni-Body   
11024  VOLVO      2011          MPV         AWD   Uni-Body   
11025  VOLVO      2011          MPV         FWD   Uni-Body   
11026  VOLVO      2011          MPV         AWD   Uni-Body   
11027  VOLVO      2011          MPV         FWD   Uni-Body   

      BLIND_SPOT_DETECTION       ARS FRNT_COLLISION_WARNING  \
0                 Standard  Standard               Standard   
1                 Standard  Standard               Standard   
2                 Standard  Standard               Standard   
3                 Standard  Standard               Standard   
4                 Standard  Standard               Standard   
...                    ...       ...                    ...   
11023             Optional  Standard               Optional   
11024             Optional  Standard               Optional   
11025             Optional  Standard               Optional   
11026             Optional  Standard                    NaN   
11027             Optional  Standard                    NaN   

      LANE_DEPARTURE_WARNING CRASH_IMMINENT_BRAKE DYNAMIC_BRAKE_SUPPORT  \
0                   Standard             Standard              Standard   
1                   Standard             Standard              Standard   
2                   Standard             Standard              Standard   
3                   Standard             Standard              Standard   
4                   Standard             Standard              Standard   
...                      ...                  ...                   ...   
11023               Optional                  NaN                   NaN   
11024               Optional                  NaN                   NaN   
11025               Optional                  NaN                   NaN   
11026                    NaN                  NaN                   NaN   
11027                    NaN                  NaN                   NaN   

       OVERALL_STARS BACKUP_CAMERA  
0                5.0           NaN  
1                NaN           NaN  
2                NaN           NaN  
3                5.0           NaN  
4                5.0           NaN  
...              ...           ...  
11023            5.0           NaN  
11024            NaN           NaN  
11025            NaN           NaN  
11026            NaN           NaN  
11027            NaN           NaN  

[11028 rows x 13 columns]

<p>Upon going over the dataset, I decided to only keep these columns. I'm only interested in the overall 5-star rating of the vehicles and seeing the correlations between the amount of safety features, the frame type, and the vehicle types. </p>

### Run the info function to check for null values

In [12]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11028 entries, 0 to 11027
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   MAKE                    11028 non-null  object 
 1   MODEL_YR                11028 non-null  int64  
 2   VEHICLE_TYPE            11028 non-null  object 
 3   DRIVE_TRAIN             11027 non-null  object 
 4   BODY_FRAME              11025 non-null  object 
 5   BLIND_SPOT_DETECTION    7696 non-null   object 
 6   ARS                     8284 non-null   object 
 7   FRNT_COLLISION_WARNING  7345 non-null   object 
 8   LANE_DEPARTURE_WARNING  6775 non-null   object 
 9   CRASH_IMMINENT_BRAKE    6419 non-null   object 
 10  DYNAMIC_BRAKE_SUPPORT   6509 non-null   object 
 11  OVERALL_STARS           4670 non-null   float64
 12  BACKUP_CAMERA           2798 non-null   object 
dtypes: float64(1), int64(1), object(11)
memory usage: 1.1+ MB


<br>
<br>
Since the overall goal of this project is to check for overall 5-Star Ratings, the rows without a rating will be dropped. With 4670 non-null values remaining, this will still be a good sample size to get reliable data.

In [13]:
data_df = data_df.dropna(subset=['OVERALL_STARS'])
data_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4670 entries, 0 to 11023
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   MAKE                    4670 non-null   object 
 1   MODEL_YR                4670 non-null   int64  
 2   VEHICLE_TYPE            4670 non-null   object 
 3   DRIVE_TRAIN             4670 non-null   object 
 4   BODY_FRAME              4669 non-null   object 
 5   BLIND_SPOT_DETECTION    3368 non-null   object 
 6   ARS                     3576 non-null   object 
 7   FRNT_COLLISION_WARNING  3346 non-null   object 
 8   LANE_DEPARTURE_WARNING  3193 non-null   object 
 9   CRASH_IMMINENT_BRAKE    2867 non-null   object 
 10  DYNAMIC_BRAKE_SUPPORT   2732 non-null   object 
 11  OVERALL_STARS           4670 non-null   float64
 12  BACKUP_CAMERA           1377 non-null   object 
dtypes: float64(1), int64(1), object(11)
memory usage: 510.8+ KB


<br>
<br>
Next will be filling in the non-null values accordingly. Going over thesee columns, filling them in as Not Available would make the most sense. 

In [14]:
data_df = data_df.fillna('Not Available')
data_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4670 entries, 0 to 11023
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   MAKE                    4670 non-null   object 
 1   MODEL_YR                4670 non-null   int64  
 2   VEHICLE_TYPE            4670 non-null   object 
 3   DRIVE_TRAIN             4670 non-null   object 
 4   BODY_FRAME              4670 non-null   object 
 5   BLIND_SPOT_DETECTION    4670 non-null   object 
 6   ARS                     4670 non-null   object 
 7   FRNT_COLLISION_WARNING  4670 non-null   object 
 8   LANE_DEPARTURE_WARNING  4670 non-null   object 
 9   CRASH_IMMINENT_BRAKE    4670 non-null   object 
 10  DYNAMIC_BRAKE_SUPPORT   4670 non-null   object 
 11  OVERALL_STARS           4670 non-null   float64
 12  BACKUP_CAMERA           4670 non-null   object 
dtypes: float64(1), int64(1), object(11)
memory usage: 510.8+ KB


In [15]:
data_df

MAKE  MODEL_YR VEHICLE_TYPE DRIVE_TRAIN BODY_FRAME  \
0           ACURA      2024           PC         FWD              
3           ACURA      2024          MPV         FWD              
4           ACURA      2024          MPV         AWD              
5           ACURA      2024          MPV         FWD              
6           ACURA      2024           PC         AWD              
...           ...       ...          ...         ...        ...   
10995      TOYOTA      2011          MPV         AWD   Uni-body   
10996      TOYOTA      2011          MPV         FWD   Uni-body   
11007  VOLKSWAGEN      2011           PC         FWD   Uni-Body   
11022       VOLVO      2011          MPV         AWD   Uni-Body   
11023       VOLVO      2011          MPV         FWD   Uni-Body   

      BLIND_SPOT_DETECTION       ARS FRNT_COLLISION_WARNING  \
0                 Standard  Standard               Standard   
3                 Standard  Standard               Standard   
4                 Standard  Standard               Standard   
5                 Standard  Standard               Standard   
6                 Optional  Standard               Standard   
...                    ...       ...                    ...   
10995        Not Available  Standard          Not Available   
10996        Not Available  Standard          Not Available   
11007        Not Available  Standard          Not Available   
11022             Optional  Standard               Optional   
11023             Optional  Standard               Optional   

      LANE_DEPARTURE_WARNING CRASH_IMMINENT_BRAKE DYNAMIC_BRAKE_SUPPORT  \
0                   Standard             Standard              Standard   
3                   Standard             Standard              Standard   
4                   Standard             Standard              Standard   
5                   Standard             Standard              Standard   
6                   Standard             Standard              Standard   
...                      ...                  ...                   ...   
10995          Not Available        Not Available         Not Available   
10996          Not Available        Not Available         Not Available   
11007          Not Available        Not Available         Not Available   
11022               Optional        Not Available         Not Available   
11023               Optional        Not Available         Not Available   

       OVERALL_STARS  BACKUP_CAMERA  
0                5.0  Not Available  
3                5.0  Not Available  
4                5.0  Not Available  
5                5.0  Not Available  
6                5.0  Not Available  
...              ...            ...  
10995            4.0  Not Available  
10996            4.0  Not Available  
11007            4.0  Not Available  
11022            5.0  Not Available  
11023            5.0  Not Available  

[4670 rows x 13 columns]

<br>
<br>
Upon display the dataframe again, I noticed that this data still has blank spaces underneath the BODY_FRAME column. So I will run the value_counts function to see if the values turn up as a blank space. 

In [34]:
data_df['BODY_FRAME'].value_counts()

BODY_FRAME
Uni-Body         2867
Frame-Based       807
Not Available     546
uni-body          264
body-on-frame     147
Uni-body           33
Body on frame       6
Name: count, dtype: int64

In [35]:
data_df = data_df.replace(r'^\s*$', 'Available' , regex=True)

In [38]:
data_df['BACKUP_CAMERA'].value_counts()

BACKUP_CAMERA
Not Available    3293
Optional          671
Standard          654
Yes                51
No                  1
Name: count, dtype: int64